unet_models.py

```diff
@@ -294,3 +299,90 @@ def forward(self, x):
         dec1 = self.dec1(torch.cat([dec2, conv1], 1))
 
         return self.final(dec1)
+
+
+class UNetResNet(nn.Module):
+    """
+        UNet (https://arxiv.org/abs/1505.04597) with Resnet34(https://arxiv.org/abs/1512.03385) encoder
+
+        Proposed by Alexander Buslaev: https://www.linkedin.com/in/al-buslaev/
+
+        """
+
+    def __init__(self, encoder_depth=34, num_classes=1, num_filters=32, dropout_2d=0.2,
+                 pretrained=False, is_deconv=False):
+        """
+        :param num_classes:
+        :param num_filters:
+        :param pretrained:
+            False - no pre-trained network is used
+            True  - encoder is pre-trained with resnet34
+        :is_deconv:
+            False: bilinear interpolation is used in decoder
+            True: deconvolution is used in decoder
+        """
+        super().__init__()
+        self.num_classes = num_classes
+        self.dropout_2d = dropout_2d
+
+        if encoder_depth == 34:
+            self.encoder = torchvision.models.resnet34(pretrained=pretrained)
+            bottom_channel_nr = 512
+        elif encoder_depth == 101:
+            self.encoder = torchvision.models.resnet101(pretrained=pretrained)
+            bottom_channel_nr = 2048
+        elif encoder_depth == 152:
+            self.encoder = torchvision.models.resnet152(pretrained=pretrained)
+            bottom_channel_nr = 2048
+        else:
+            raise NotImplementedError('only 34, 101, 152 version of Resnet are implemented')
+
+        self.pool = nn.MaxPool2d(2, 2)
+
+        self.relu = nn.ReLU(inplace=True)
+
+        self.conv1 = nn.Sequential(self.encoder.conv1,
+                                   self.encoder.bn1,
+                                   self.encoder.relu,
+                                   self.pool)
+
+        self.conv2 = self.encoder.layer1
+
+        self.conv3 = self.encoder.layer2
+
+        self.conv4 = self.encoder.layer3
+
+        self.conv5 = self.encoder.layer4
+
+        self.center = DecoderBlockV2(bottom_channel_nr, num_filters * 8 * 2, num_filters * 8, is_deconv)
+        self.dec5 = DecoderBlockV2(bottom_channel_nr + num_filters * 8, num_filters * 8 * 2, num_filters * 8, is_deconv)
+        self.dec4 = DecoderBlockV2(int(bottom_channel_nr / 2.) + num_filters * 8, num_filters * 8 * 2, num_filters * 8,
+                                   is_deconv)
+        self.dec3 = DecoderBlockV2(int(bottom_channel_nr / 4.) + num_filters * 8, num_filters * 4 * 2, num_filters * 2,
+                                   is_deconv)
+        self.dec2 = DecoderBlockV2(int(bottom_channel_nr / 8.) + num_filters * 2, num_filters * 2 * 2,
```

Comment:
@jakubczakon it's quite a redundant comment, but maybe `bottom_channel_nr // 2` etc. would look cleaner

Reply:
@taraspiotr I think you are right, let me change that/check that it works